runtime/jcl/common/jdk_internal_misc_ScopedMemoryAccess.cpp

```diff
@@ -34,11 +36,58 @@ Java_jdk_internal_misc_ScopedMemoryAccess_registerNatives(JNIEnv *env, jclass cl
 {
 }
 
+static UDATA
+closeScope0FrameWalkFunction(J9VMThread *vmThread, J9StackWalkState *walkState)
+{
+	if (*(UDATA *)walkState->userData2) { /* scope has been found */
+		return J9_STACKWALK_STOP_ITERATING;
+	}
+	return J9_STACKWALK_KEEP_ITERATING;
+}
+
+static void
+closeScope0OSlotWalkFunction(J9VMThread *vmThread, J9StackWalkState *walkState, j9object_t *slot, const void *stackLocation)
+{
+	J9Method *method = walkState->method;
+	if (NULL != method) {
+		J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(method);
+		if (NULL != romMethod && J9ROMMETHOD_HAS_EXTENDED_MODIFIERS(romMethod)) {
+			U_32 extraModifiers = getExtendedModifiersDataFromROMMethod(romMethod);
+			if (J9ROMMETHOD_HAS_SCOPED_ANNOTATION(extraModifiers)) {
+				if (slot == walkState->userData1) {
+					*(UDATA *)walkState->userData2 = TRUE;
+				}
+			}
+		}
+	}
+}
+
 jboolean JNICALL
 Java_jdk_internal_misc_ScopedMemoryAccess_closeScope0(JNIEnv *env, jobject instance, jobject scope, jobject exception)
 {
-	Assert_JCL_unimplemented();
-	return JNI_FALSE;
+	J9VMThread *currentThread = (J9VMThread *)env;
+	J9JavaVM *vm = currentThread->javaVM;
+	UDATA scope_found = FALSE;
+	vm->internalVMFunctions->acquireExclusiveVMAccess(currentThread);
+
+	J9VMThread *walkThread = J9_LINKED_LIST_START_DO(vm->mainThread);
+	while (walkThread != NULL) {
+		J9StackWalkState walkState;
+		walkState.walkThread = walkThread;
+		walkState.flags = J9_STACKWALK_ITERATE_FRAMES | J9_STACKWALK_ITERATE_O_SLOTS;
+		walkState.skipCount = 0;
+		walkState.userData1 = (void *)scope;
+		walkState.userData2 = (void *)&scope_found;
+		walkState.frameWalkFunction = closeScope0FrameWalkFunction;
+		walkState.objectSlotWalkFunction = closeScope0OSlotWalkFunction;
+
+		vm->walkStackFrames(walkThread, &walkState);
+
+		walkThread = J9_LINKED_LIST_NEXT_DO(vm->mainThread, walkThread);
```

Comment:
Check if scope found after `walkStackFrames` and break so don't have to continue walking rest of the threads